Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,528 @@
+      // NOTE: This line used to be before we explicitly fail the status promise but it was exaserbating a race condition and causing crashes. See https://bugs.swift.org/browse/SR-15108.
+      if let task = self.task {
+        task.cancel()
+      }
+
+    case .completed:
+      ()
+    }
+  }
+
+  @inlinable
+  internal func sendInterceptedPart(
+    _ part: GRPCServerResponsePart<Response>,
+    promise: EventLoopPromise<Void>?
+  ) {
+    switch part {
+    case let .metadata(headers):
+      self.context.responseWriter.sendMetadata(headers, flush: true, promise: promise)
+
+    case let .message(message, metadata):
+      do {
+        let bytes = try self.serializer.serialize(message, allocator: ByteBufferAllocator())
+        self.context.responseWriter.sendMessage(bytes, metadata: metadata, promise: promise)
+      } catch {
+        // Serialization failed: fail the promise and send end.
+        promise?.fail(error)
+        let (status, trailers) = ServerErrorProcessor.processLibraryError(
+          error,
+          delegate: self.context.errorDelegate
+        )
+        // Loop back via the interceptors.
+        self.interceptors.send(.end(status, trailers), promise: nil)
+      }
+
+    case let .end(status, trailers):
+      self.context.responseWriter.sendEnd(status: status, trailers: trailers, promise: promise)
+    }
+  }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension GRPCAsyncServerHandler {
+  /// Async-await wrapper for `interceptResponse(_:metadata:promise:)`.
+  ///
+  /// This will take care of ensuring it executes on the right event loop.
+  @inlinable
+  internal func interceptResponse(
+    _ response: Response,
+    metadata: MessageMetadata
+  ) async throws {
+    let promise = self.context.eventLoop.makePromise(of: Void.self)
+    if self.context.eventLoop.inEventLoop {
+      self.interceptResponse(response, metadata: metadata, promise: promise)
+    } else {
+      self.context.eventLoop.execute {
+        self.interceptResponse(response, metadata: metadata, promise: promise)
+      }
+    }
+    try await promise.futureResult.get()
+  }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension GRPCAsyncServerHandler {
+  @inlinable
+  public convenience init(
+    context: CallHandlerContext,
+    requestDeserializer: Deserializer,
+    responseSerializer: Serializer,
+    interceptors: [ServerInterceptor<Request, Response>],
+    wrapping unary: @escaping @Sendable(Request, GRPCAsyncServerCallContext) async throws
+      -> Response
+  ) {
+    self.init(
+      context: context,
+      requestDeserializer: requestDeserializer,
+      responseSerializer: responseSerializer,
+      interceptors: interceptors,
+      observer: { requestStream, responseStreamWriter, context in
+        for try await request in requestStream.prefix(1) {
+          let response = try await unary(request, context)
+          try await responseStreamWriter.sendResponse(response)
+        }
```

Comment:
I think with the current implementation if there are no request messages we'll still finish the RPC with `.ok`. I think we need to throw an appropriate status error here if that's the case.

Reply:
I've added some code to guard against this in 71d85d57 but I don't want to mark this as resolved because I want to confirm that this is what you were expecting to see?